taskchampion/taskchampion/src/server/cloud/server.rs

```diff
@@ -0,0 +1,1130 @@
+/// All other objects are invalid and not visible outside this type.
+///
+/// Snapshots are stored as objects with name `s-VERSION` where `VERSION` is the version at which
+/// the snapshot was made. These objects are created with simple `put` requests, as any snapshot
+/// for a given version is interchangeable with any other.
+///
+/// ## Cleanup
+///
+/// Cleanup of unnecessary data is performed probabalistically after `add_version`, although any
+/// errors are ignored.
+///
+///  - Any versions not reachable from "latest" and which cannot become "latest" are deleted.
+///  - Any snapshots older than the most recent are deleted.
+///  - Any versions older than [`MAX_VERSION_AGE_SECS`] which are incorporated into a snapshot
+///    are deleted.
+pub(in crate::server) struct CloudServer<SVC: Service> {
+    service: SVC,
+
+    /// The Cryptor supporting encryption and decryption of objects in this server.
+    cryptor: Cryptor,
+
+    /// The probability (0..255) that this run will perform cleanup.
+    cleanup_probability: u8,
+
+    /// For testing, a function that is called in the middle of `add_version` to simulate
+    /// a concurrent change in the service.
+    #[cfg(test)]
+    add_version_intercept: Option<fn(service: &mut SVC)>,
+}
+
+const LATEST: &[u8] = b"latest";
+const DEFAULT_CLEANUP_PROBABILITY: u8 = 13; // about 5%
+
+#[cfg(not(test))]
+const MAX_VERSION_AGE_SECS: u64 = 3600 * 24 * 180; // about half a year
+
+fn version_to_bytes(v: VersionId) -> Vec<u8> {
+    v.as_simple().to_string().into_bytes()
+}
+
+impl<SVC: Service> CloudServer<SVC> {
+    pub(in crate::server) fn new(service: SVC, encryption_secret: Vec<u8>) -> Result<Self> {
+        let cryptor = Cryptor::new(b"Taskchampion", &encryption_secret.into())?;
+        Ok(Self {
+            service,
+            cryptor,
+            cleanup_probability: DEFAULT_CLEANUP_PROBABILITY,
+            #[cfg(test)]
+            add_version_intercept: None,
+        })
+    }
+
+    /// Generate an object name for the given parent and child versions.
+    fn version_name(parent_version_id: &VersionId, child_version_id: &VersionId) -> Vec<u8> {
+        format!(
+            "v-{}-{}",
+            parent_version_id.as_simple(),
+            child_version_id.as_simple()
+        )
+        .into_bytes()
+    }
+
+    /// Parse a version name as generated by `version_name`.
+    fn parse_version_name(name: &[u8]) -> Option<(VersionId, VersionId)> {
+        if name.len() != 2 + 32 + 1 + 32 || !name.starts_with(b"v-") || name[2 + 32] != b'-' {
+            return None;
+        }
+        let Ok(parent_version_id) = VersionId::try_parse_ascii(&name[2..2 + 32]) else {
+            return None;
+        };
+        let Ok(child_version_id) = VersionId::try_parse_ascii(&name[2 + 32 + 1..]) else {
+            return None;
+        };
+        Some((parent_version_id, child_version_id))
+    }
```

Comment:
This might happen if the user adds other files to the bucket (which they shouldn't!). If that generates a warning, users will see a lot of them! Also, we don't have a way to log user-visible warnings here.  `log::warn!` won't appear unless the log level is set properly, which is currently impossible (#3220). I'll add a comment as to when None is returned here.